File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(int k, vector<vector<int>> dungeons) {
    int answer = -1;
    
    vector<int> arr;
    
    int len = dungeons.size();
    
    for(int i=1; i<=len ;i++) 
    {
        arr.push_back(i);
    }
    
    do 
    {
        
        int cnt =0;
        int hp = k;
        
        for(auto& idx : arr)
        {
            auto& item = dungeons[idx-1];
            
            if(item[0] > hp) break;
            
            hp -= item[1];
            cnt +=1;
        }
        if(cnt > answer) answer = cnt;
    }
    while(next_permutation(arr.begin(), arr.end()));
    
    
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    var answer = 0;
    
    const len = dungeons.length;
    for(let i=0; i< len ; i++) {
        if(k >= dungeons[i][0]) {
            let arr = [
                ...dungeons.slice(0, i),
                ...dungeons.slice(i+1, len)
            ]; 
            
            const res = solution(k- dungeons[i][1], arr) +1;
            
            if(res > answer) answer = res;
        }
    }
    
    return answer;
}